File: 1100-find-k-length-substrings-with-no-repeated-characters/1100-find-k-length-substrings-with-no-repeated-characters.cpp
class Solution {
public:
    int numKLenSubstrNoRepeats(string s, int k) {
        if ( s.size() < k )
            return 0;
        
        unordered_map<char, int> hmap;
        int gCount = 0;
        
        // Initiating the first substring of leght k
        for (int i = 0; i < k; i++) {
            if (hmap.find(s[i]) == hmap.end())
                hmap[s[i]] = 1;
            else
                hmap[s[i]] += 1;
        }
        
        if (hmap.size() == k)
            gCount = 1;
        
        for (int i = k; i < s.size(); i++) {
            if (hmap.find(s[i]) == hmap.end())
                hmap[s[i]] = 1;
            else
                hmap[s[i]] += 1;
            
            // decrement/erase the otugoing char
            hmap[s[i-k]] -= 1;
            if (hmap[s[i - k]] == 0)
                hmap.erase(s[i-k]);
            
            if (hmap.size() == k)
                gCount += 1;
        } 
        
        return gCount;
    }
};

File: 1100-find-k-length-substrings-with-no-repeated-characters/1100-find-k-length-substrings-with-no-repeated-characters.py
class Solution(object):
    def numKLenSubstrNoRepeats(self, s, k):
        """
        :type s: str
        :type k: int
        :rtype: int
        """
        if k > len(s):
            return 0
        
        hmap = {}
        
        for i in range(k):
            if s[i] in hmap:
                hmap[s[i]] += 1
            else: 
                hmap[s[i]] = 1
        
        if len(hmap) == k:
            strcount = 1
        else:
            strcount = 0
        
        for i in range(k, len(s)):
            if s[i] in hmap:
                hmap[s[i]] += 1
            else:
                hmap[s[i]] = 1
            
            hmap[s[i-k]] -= 1
            if hmap[s[i-k]] == 0:
                del hmap[s[i-k]]
            
            if len(hmap) == k:
                strcount += 1
       
        return strcount
            
            